### PR TITLE
chore(api): prevent API leaks (cache and thread)

### PR DESCRIPTION
**Pull request by kisvegabor**
There were some (at least in my opinion) API leaks. Meaning some modules were exposes which shouldn't be available for the user by default:
- cache: I think it's clear that it's not for the user
- draw API in general: use the widgets instead, custom drawing is an advanced thing
- os abstraction: it's an internal LVGL thing, if the user needs it it can be included directly

What do you think about it?

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Review by liamHowatt — COMMENTED** [review]
Amazing! Thank you for taking care of this.

Note the CI failures.

- `lv_windows_context.c` needs to include `lv_os.h` to get `lv_lock`
- [This file](https://github.com/lvgl/lv_binding_micropython/blob/96d6d3538fec4de99c200c1bb388b02b95cbbdaf/driver/esp32/espidf.c) needs to include `lv_draw_sw.h` to get `lv_draw_sw_rgb565_swap`.

Or they could be solved by making/keeping them public.

**Comment by kisvegabor** [comment]
Both should be fixed. For `lv_draw_sw_rgb565_swap` I added `lv_draw_sw_utils.c/h` which are public.

**Comment by liamHowatt** [comment]
Good idea! Fyi it looks like you forgot to save/commit the new `lv_draw_sw_utils.c/h` files.

**Comment by kisvegabor** [comment]
Upps, true. :slightly_smiling_face:

File: src/stdlib/builtin/lv_tlsf.h [hunk 1]
```diff
@@ -41,7 +41,7 @@
 ** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-#include "../../osal/lv_os.h"
+//#include "../../osal/lv_os.h"
```
Comment (liamHowatt) [inline src/stdlib/builtin/lv_tlsf.h]:
```suggestion
```

File: src/draw/sw/lv_draw_sw_utils.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/***********************
+ * GLOBAL VARIABLES
+ ***********************/
+
+/**********************
+ *      MACROS
+ **********************/
+
+
+#endif /*LV_USE_DRAW_SW_UTILS*/
```
Comment (liamHowatt) [inline src/draw/sw/lv_draw_sw_utils.h]:
```suggestion
#endif /*LV_USE_DRAW_SW*/
```

**Review by liamHowatt — COMMENTED** [review]
Amazing!

**Comment by kisvegabor** [comment]
Thanks, both are fixed.

**Review by liamHowatt — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Review by W-Mai — APPROVED** [review]

**Comment by XuNeo** [comment]
> draw API in general: use the widgets instead, custom drawing is an advanced thing

Agreed. The `lv_demo_xxx.c` that includes `lvgl_private.h` should be updated to a separate source file like `widget_spectrum.c` for lv_demo_music.

It indicates the lvgl_private.h is for advanced custom drawing and not required for general application, aka the demo.

**Comment by kisvegabor** [comment]
@XuNeo It makes send to create a new file for the spectrum widget. Let's address it separately.

**Comment by kdschlosser** [comment]
when including header files into C files there is no risk to polluting the public API. it's only when there are private header files that are included into public ones is where the issue stems from. If there is a specific piece like a structure that is used in a function call that is in a private header file then consider 2 things making the function private or forward declaring the structure in the public header file instead of including the private header file.

IMHO If a function has a parameter that is typed with something that is private then the function itself should also be made private. The exceptions to this are the core structures like `lv_obj_t`, `lv_display_t`, etc... Something like `lv_cache_t` where the user would not be able to collect that structure from anywhere to even be able to pass it to the function renders the function completely useless. Where this becomes an issue is with high level programming language bindings because the linker is going to include the code due to how bindings work or when compiling LVGL as a shared library. It's wasted resources for something that ultimately is not able to be used for anything.

The state of the include tree is also pretty messy and needs a little TLC. If the include tree isn't kept clean it is hard to figure out how the public API is getting contaminated with private API declarations. That is the reason why I am bringing this up here.

I have spent some time hammering out the what the include tree should look like.
This is what I have so far with what should be included in the different header files. This is for the local includes only and it was made prior to this PR getting merged so it may need to be adjusted for changes made in the PR. The files listed are checked 100%. the rest of the files I am still working on. This is ONLY the header files. I have not gone through the source files at all.

```
file: src/core/lv_group.h
includes: src/misc/lv_types.h

file: src/core/lv_obj_class.h
includes: src/misc/lv_types.h

file: src/core/lv_obj_draw.h
includes: src/misc/lv_types_h
          src/draw/lv_draw_rect.h
          src/draw/lv_draw_label.h
          src/draw/lv_draw_image.h
          src/draw/lv_draw_line.h
          src/draw/lv_draw_arc.h

file: src/core/lv_obj_event.h
includes: src/misc/lv_types.h
          src/misc/lv_event.h
          src/misc/lv_area.h
              
file: src/core/lv_obj.h
includes: src/misc/lv_types.h
          src/core/lv_obj_property.h
          src/core/lv_group.h

file: src/core/lv_obj_pos.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h

file: src/core/lv_obj_property.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_color.h
          src/misc/lv_style.h

file: src/core/lv_obj_scroll.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_anim.h

file: src/core/lv_obj_style_gen.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/core/lv_obj_style.h
          src/misc/lv_style.h
          src/misc/lv_text.h
          src/misc/lv_color.h
          src/misc/lv_color_op.h
          src/misc/lv_bidi.h
          src/layouts/flex/lv_flex.h
          src/layouts/grid/lv_grid.h

file: src/core/lv_obj_style.h
includes: src/misc/lv_types.h
          src/misc/lv_style.h
          src/misc/lv_text.h

file: src/core/lv_obj_tree.h
includes: src/misc/lv_types.h

file: src/core/lv_refr.h
includes: src/misc/lv_types.h

file: src/display/lv_display.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_event.h
          src/misc/lv_color.h

file: src/draw/lv_draw_arc.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_color.h
          src/draw/lv_draw.h
              
file: src/draw/lv_draw_buf.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_color.h
          src/misc/lv_math.h
          src/draw/lv_image_dsc.h

file: src/draw/lv_draw.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_color.h
          
file: src/draw/lv_draw_image.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_color.h
          src/misc/lv_style.h
          src/draw/lv_image_dsc.h
          src/draw/lv_draw.h
          src/draw/lv_image_decoder.h
          src/draw/lv_image_dsc.h

file: src/draw/lv_draw_label.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_bidi.h
          src/misc/lv_style.h
          src/misc/lv_text.h
          src/misc/lv_color.h
          src/draw/lv_draw.h
          src/draw/lv_draw_rect.h
          
file: src/draw/lv_draw_line.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_style.h
          src/misc/lv_color.h
          src/draw/lv_draw.h
          
file: src/draw/lv_draw_mask.h
includes: src/misc/lv_types.h
    
file: src/draw/lv_draw_rect.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_style.h
          src/misc/lv_color.h
          src/draw/lv_draw.h

file: src/draw/lv_draw_triangle.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/misc/lv_style.h
          src/misc/lv_color.h
          src/draw/lv_draw.h

file: src/draw/lv_image_decoder.h
includes: src/misc/lv_types.h
          src/misc/lv_area.h
          src/draw/lv_image_dsc.h   
            
file: src/draw/lv_image_dsc.h
includes: None

file: src/font/lv_binfont_loader.h
includes: src/misc/lv_types.h
 
file: src/font/lv_font_fmt_txt.h
includes: src/misc/lv_types.h      
          src/font/lv_font.h

file: src/font/lv_font.h
includes: src/misc/lv_types.h      

file: src/font/lv_symbol_def.h
includes: None

file: src/indev/lv_indev.h
includes: src/misc/lv_types.h      
          src/misc/lv_event.h
          src/misc/lv_area.h

file: src/indev/lv_indev_scroll.h
includes: src/misc/lv_types.h      
          src/misc/lv_area.h        

file: src/layouts/flex/lv_flex.h
includes: src/misc/lv_types.h      

file: src/layouts/grid/lv_grid.h
includes: src/misc/lv_types.h    
          src/misc/lv_area.h        

file: src/misc/lv_anim.h
includes: src/misc/lv_types.h    
          src/misc/lv_math.h    

file: src/misc/lv_anim_timeline.h
includes: src/misc/lv_anim.h    

file: src/misc/lv_area.h
includes: src/misc/lv_types.h    
          src/misc/lv_math.h    

file: src/misc/lv_array.h
includes: src/misc/lv_types.h    

file: src/misc/lv_async.h
includes: src/misc/lv_types.h    

file: src/misc/lv_bidi.h
includes: src/misc/lv_text.h    

file: src/misc/lv_color.h
includes: src/misc/lv_types.h    

file: src/misc/lv_color_op.h
includes: src/misc/lv_types.h    
          src/misc/lv_color.h  

file: src/misc/lv_event.h
includes: src/misc/lv_types.h    
          src/misc/lv_array.h  

file: src/misc/lv_fs.h
includes: src/misc/lv_types.h    

file: src/misc/lv_ll.h
includes: None 

file: src/misc/lv_log.h
includes: None 

file: src/misc/lv_math.h
includes: None 

file: src/misc/lv_style_gen.h
includes: src/misc/lv_types.h    
          src/misc/lv_area.h  
          src/misc/lv_style.h  
          src/misc/lv_text.h  
          src/misc/lv_color.h  
          src/misc/lv_color_op.h  
          src/misc/lv_bidi.h  
          src/layouts/flex/lv_flex.h  
          src/layouts/grid/lv_grid.h

file: src/misc/lv_style.h
includes: src/misc/lv_types.h    
          src/misc/lv_area.h  
          src/misc/lv_color.h  
          src/misc/lv_anim.h  

file: src/misc/lv_text.h
includes: src/misc/lv_types.h    
          src/misc/lv_area.h  

file: src/misc/lv_types.h
includes: None  
```

**Pull request merged**